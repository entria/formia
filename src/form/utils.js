import get from 'lodash.get';
import set from 'lodash.set';
import isEqual from 'lodash.isequal';

export const handleInitialValuesUpdate = (
  currentValues = {},
  currentInitialValues = {},
  newInitialValues = {},
) => {
  const updatedValues = { ...currentValues };

  Object.keys(newInitialValues).map(fieldName => {
    const fieldCurrentValue = get(currentValues, fieldName, null);
    const fieldCurrentInitialValue = get(currentInitialValues, fieldName, null);
    const isFieldPristine = isEqual(fieldCurrentValue, fieldCurrentInitialValue);

    if (!isFieldPristine) {
      return;
    }

    const fieldNewInitialValue = get(newInitialValues, fieldName, null);
    set(updatedValues, fieldName, fieldNewInitialValue);
  });

  return updatedValues;
};
