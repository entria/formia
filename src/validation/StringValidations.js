// @flow
import { Strings } from '@entria/utils';

import type { RuleError } from './Validation';

export const isAlpha = () => (value: any): ?RuleError => {
  if (Strings.isEmpty(value)) {
    return null;
  }

  if (Strings.isAlpha(value)) {
    return null;
  }

  return {
    code: 'Validation.Strings.isAlpha',
    message: 'Only alpha characters are allowed',
    params: [],
  };
};

export const isEmail = () => (value: any): ?RuleError => {
  if (Strings.isEmpty(value)) {
    return null;
  }

  if (Strings.isEmail(value)) {
    return null;
  }

  return {
    code: 'Validation.Strings.isEmail',
    message: 'Invalid email address',
    params: [],
  };
};

export const minLength = (quantity: number) => (value: any): ?RuleError => {
  if (Strings.isEmpty(value)) {
    return null;
  }

  const isValid = value.length >= quantity;
  if (isValid) {
    return null;
  }

  return {
    code: 'Validation.Strings.minLength',
    message: `Must have at least ${quantity} characters`,
    params: [quantity],
  };
};

export const maxLength = (quantity: number) => (value: any): ?RuleError => {
  if (Strings.isEmpty(value)) {
    return null;
  }

  const isValid = value.length <= quantity;
  if (isValid) {
    return null;
  }

  return {
    code: 'Validation.Strings.maxLength',
    message: `Must have at most ${quantity} characters`,
    params: [quantity],
  };
};

export const minWords = (quantity: number) => (value: any): ?RuleError => {
  if (Strings.isEmpty(value)) {
    return null;
  }

  const words = value ? value.trim().split(' ') : '';
  const isValid = words.length >= quantity;
  if (isValid) {
    return null;
  }

  return {
    code: 'Validation.Strings.minWords',
    message: `Must have at least ${quantity} words`,
    params: [quantity],
  };
};

export const maxWords = (quantity: number) => (value: any): ?RuleError => {
  if (Strings.isEmpty(value)) {
    return null;
  }

  const words = value ? value.trim().split(' ') : '';
  const isValid = words.length <= quantity;
  if (isValid) {
    return null;
  }

  return {
    code: 'Validation.Strings.maxWords',
    message: `Must have at most ${quantity} words`,
    params: [quantity],
  };
};
