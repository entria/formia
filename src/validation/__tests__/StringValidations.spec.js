import * as Strings from '../StringValidations';

import { setupTest } from '../../../test/utils';

beforeEach(async () => setupTest());

it('isAlpha - empty', () => {
  expect(Strings.isAlpha()()).toMatchSnapshot();
});

it('isAlpha - valid alpha', () => {
  expect(Strings.isAlpha()('John Doe')).toMatchSnapshot();
});

it('isAlpha - invalid alpha', () => {
  expect(Strings.isAlpha()('John Doe #@*&')).toMatchSnapshot();
});

it('isEmail - empty', () => {
  expect(Strings.isEmail()()).toMatchSnapshot();
});

it('isEmail - valid email', () => {
  expect(Strings.isEmail()('john@doe.com')).toMatchSnapshot();
});

it('isEmail - invalid email', () => {
  expect(Strings.isEmail()('fake@d.c.b')).toMatchSnapshot();
});

it('minLength - empty', () => {
  expect(Strings.minLength(10)()).toMatchSnapshot();
});

it('minLength - valid sentence', () => {
  expect(Strings.minLength(5)('12345')).toMatchSnapshot();
});

it('minLength - invalid sentence', () => {
  expect(Strings.minLength(5)('123')).toMatchSnapshot();
});

it('maxLength - empty', () => {
  expect(Strings.maxLength(10)()).toMatchSnapshot();
});

it('maxLength - valid sentence', () => {
  expect(Strings.maxLength(5)('12345')).toMatchSnapshot();
});

it('maxLength - invalid sentence', () => {
  expect(Strings.maxLength(5)('123456')).toMatchSnapshot();
});

it('minWords - empty', () => {
  expect(Strings.minWords(2)()).toMatchSnapshot();
});

it('minWords - valid sentence', () => {
  expect(Strings.minWords(2)('John Doe')).toMatchSnapshot();
});

it('minWords - invalid sentence', () => {
  expect(Strings.minWords(2)('John')).toMatchSnapshot();
});

it('maxWords - empty', () => {
  expect(Strings.maxWords(2)()).toMatchSnapshot();
});

it('maxWords - valid sentence', () => {
  expect(Strings.maxWords(2)('John Doe')).toMatchSnapshot();
});

it('maxWords - invalid sentence', () => {
  expect(Strings.maxWords(2)('Mr. John Doe')).toMatchSnapshot();
});
