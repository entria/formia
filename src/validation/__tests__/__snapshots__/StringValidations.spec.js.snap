// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isAlpha - empty 1`] = `null`;

exports[`isAlpha - invalid alpha 1`] = `
Object {
  "code": "Validation.Strings.isAlpha",
  "message": "Only alpha characters are allowed",
  "params": Array [],
}
`;

exports[`isAlpha - valid alpha 1`] = `null`;

exports[`isEmail - empty 1`] = `null`;

exports[`isEmail - invalid email 1`] = `
Object {
  "code": "Validation.Strings.isEmail",
  "message": "Invalid email address",
  "params": Array [],
}
`;

exports[`isEmail - valid email 1`] = `null`;

exports[`minLength - empty 1`] = `null`;

exports[`minLength - invalid sentence 1`] = `
Object {
  "code": "Validation.Strings.minLength",
  "message": "Must have at least 5 characters",
  "params": Array [
    5,
  ],
}
`;

exports[`minLength - valid sentence 1`] = `null`;

exports[`minWords - empty 1`] = `null`;

exports[`minWords - invalid sentence 1`] = `
Object {
  "code": "Validation.Strings.minWords",
  "message": "Must have at least 2 words",
  "params": Array [
    2,
  ],
}
`;

exports[`minWords - valid sentence 1`] = `null`;
